### DISCUSSION TITLE
CoCo/CoCo3 - Request for Game Master Cartridge (GMC) support

### DISCUSSION THREAD

**Discussion by mrgw454** [spotlessmind1975/ugbasic#718] [Ideas]
Hi Marco,

There is an audio cartridge available for all models of CoCo that provides much better sound capability than the stock DAC's alone.  It uses a period correct SN76489 IC.  In addition, the cartridge provides bank switching capability (up to 64K via a ROM socket).  I'm mainly interested in the audio capabilities, personally.  There has been discussion of adding enhanced PLAY commands to ECB/DECB, but it would seem like a natural fit to incorporate those commands/features into ugBasic.

The cartridge was designed/built by John Linville (long time CoCo enthusiast).  MAME and XRoar both have support for the GMC cart, so it's easily used in both emulators.

I realize there are more urgent/important things you are working on, but if you are interested, here's some links to additional information about this cartridge:

https://www.cocopedia.com/wiki/index.php/Game_Master_Cart
https://docs.google.com/document/d/17JzrNqIHZaFmtHEeFSevpkJtbWOn2-2EjxMB2cQlxsY/edit#heading=h.ac9evjq4qq2r
https://www.vintageisthenewold.com/game-master-cartridge-enables-better-games-for-coco

Thank you for your consideration!

**Comment by linvjw** [comment]
FWIW, there are a few other documents relevant to the GMC and it's use here:

[https://drive.google.com/drive/u/0/folders/1FWSpWshl_GJevk85hsm54b62SGGojyB1](https://drive.google.com/drive/u/0/folders/1FWSpWshl_GJevk85hsm54b62SGGojyB1)

I am generally happy to answer questions as well, but really the device isn't too complicated. The hard part is knowing what you want to make it do. :-)

**Comment by spotlessmind1975** [comment]
Hi @mrgw454 and @linvjw, and **thank you for the suggestion!**

I just added support for the audio chipset contained in the Game Master Cartridge to the `beta` ( ticket #749 ), both for `coco` and `coco3` targets. To use it, simply use standard audio commands, such as `SOUND` or `BOOM`. The best thing, however, is to play a `MIDI` or a `PSG`. They are both supported, although I only tested the PSG file format. In this case the music is played in the background, while the rest of the program (including multitasking) runs without interruptions or interference. In the case of the `coco3`, the music can be stored in one of the GIME banks (maximum 8 KB per song), so as not to occupy the main memory.

In any case, the syntax for loading and playing music in MIDI format is as follows:

     soundtrack := LOAD MUSIC("track.mid")
     MUSIC soundtrack

For music in PSG format it is the following:

     soundtrack := LOAD("track.psg")
     MUSIC soundtrack PSG

For `coco3` only, you can load music into one of the GIME banks with the command:

     soundtrack := LOAD("track.psg") BANKED
     MUSIC soundtrack PSG

Click here for more information:
https://ugbasic.iwashere.eu/manual/audio
https://ugbasic.iwashere.eu/target/coco#audio
https://ugbasic.iwashere.eu/target/coco3#audio

That's it, I think that's it - let me know if you have any other ideas. :)

Thank you again and happy retrocoding! ^_^

**Reply by mrgw454** [reply]
Hi Marco,

This is great news and I certainly appreciate you taking the time to add this support.  I'll be trying to this out (as soon I pull and compile the latest version of ugBasic from your repo).  

Thank you again!

-Ron
